SpockOpaleye/src/DataTypes.hs

```diff
@@ -16,51 +47,78 @@ newtype TenantId = TenantId Int
+  , _tenantpolyUpdatedat        :: updated_at
+  , _tenantpolyName             :: name
+  , _tenantpolyFirstname        :: fname
+  , _tenantpolyLastname         :: lname
+  , _tenantpolyEmail            :: email
+  , _tenantpolyPhone            :: phone
+  , _tenantpolyStatus           :: status
+  , _tenantpolyOwnerid          :: owner_id
+  , _tenantpolyBackofficedomain :: b_domain
+} deriving (Show, Generic)
+
 
-type Tenant = TenantPoly TenantId Text Text Text Text Text TenantStatus (Maybe UserId) Text
+type Tenant = TenantPoly TenantId UTCTime UTCTime Text Text Text Text Text TenantStatus (Maybe UserId) Text
 
-type TenantIncoming = TenantPoly () Text Text Text Text Text () (Maybe UserId) Text
+type TenantIncoming = TenantPoly () () () Text Text Text Text Text () (Maybe UserId) Text
+type TenantIncomingCreatable = TenantPoly () UTCTime UTCTime Text Text Text Text Text () (Maybe UserId) Text
 
 data UserStatus = UserStatusActive | UserStatusInActive | UserStatusBlocked
   deriving (Show)
 
 newtype UserId = UserId Int
   deriving (Show, Generic)
 
-data UserPoly key tenant_id username password firstname lastname status  = User {
-  user_id        :: key,
-  user_tenantid  :: tenant_id,
-  user_username  :: username,
-  user_password  :: password,
-  user_firstname :: firstname,
-  user_lastname  :: lastname,
-  user_status    :: status
-}
+data UserPoly key created_at updated_at tenant_id username password firstname lastname status  = User {
+    _userpolyId        :: key
+  , _userpolyCreatedat :: created_at
+  , _userpolyUpdatedat :: updated_at
+  , _userpolyTenantid  :: tenant_id
+  , _userpolyUsername  :: username
+  , _userpolyPassword  :: password
+  , _userpolyFirstname :: firstname
+  , _userpolyLastname  :: lastname
+  , _userpolyStatus    :: status
+} deriving (Show)
+
+type User = UserPoly UserId UTCTime UTCTime TenantId Text BcryptPassword (Maybe Text) (Maybe Text) UserStatus
 
-type User = UserPoly UserId TenantId Text BcryptPassword (Maybe Text) (Maybe Text) UserStatus
+type UserIncoming = UserPoly () () () TenantId Text Text (Maybe Text) (Maybe Text) ()
 
 data Permission = Read | Create | Update | Delete
   deriving (Show)
 
 newtype RoleId = RoleId Int
   deriving (Show)
 
-data RolePoly key tenant_id name permission = Role
-  { role_id         :: key
-  , role_tenantid   :: tenant_id
-  , role_name       :: name
-  , role_permission :: permission
-  } deriving (Show)
+data RolePoly key tenant_id name permission created_at updated_at  = Role {
+    _rolepolyId         :: key
+  , _rolepolyTenantid   :: tenant_id
+  , _rolepolyName       :: name
+  , _rolepolyPermission :: permission
+  , _rolepolyCreatedat  :: created_at
+  , _rolepolyUpdatedat  :: updated_at
+} deriving (Show)
+
+type Role = RolePoly RoleId TenantId Text (NonEmpty Permission) UTCTime UTCTime
+type RoleIncoming = RolePoly () TenantId Text (NonEmpty Permission) () ()
+
+data Auditable a = Auditable { _data:: a, _log:: Value }  deriving (Show)
```

Comment:
Are we exporting the `Auditable` constructor? Should we be exporting it? Should it be possible to extract an `x` out of an `Auditable x`? Is there anything that a user can do on `x` that he can't do on `Auditable x`? Can the programmer take an `x` and manually create an `Auditable x` out of it? Isn't that a highly unsafe operation, defeating the purpose of audit logs in the first place?